packages/kbn-config-schema/src/types/object_type.ts

```diff
@@ -26,9 +26,26 @@ export type Props = Record<string, Type<any>>;
 
 export type TypeOf<RT extends Type<any>> = RT['type'];
 
+type OptionalProperties<Base extends Props> = Pick<
+  Base,
+  {
+    [Key in keyof Base]: undefined extends TypeOf<Base[Key]> ? Key : never;
+  }[keyof Base]
+>;
+
+type RequiredProperties<Base extends Props> = Pick<
+  Base,
+  {
+    [Key in keyof Base]: undefined extends TypeOf<Base[Key]> ? never : Key;
+  }[keyof Base]
+>;
+
 // Because of https://github.com/Microsoft/TypeScript/issues/14041
 // this might not have perfect _rendering_ output, but it will be typed.
-export type ObjectResultType<P extends Props> = Readonly<{ [K in keyof P]: TypeOf<P[K]> }>;
+export type ObjectResultType<P extends Props> = Readonly<
```

Comment:
![maxresdefault](https://user-images.githubusercontent.com/3198181/79764963-523dd000-8326-11ea-9c34-94259d54201d.jpg)

Reply:
Yea, TS magic... TBH I really thought there would be an **easier** way to type that, but AFAIK there is not, as the `?` modifier for an object key(s) cannot be directly inferred / depends on the associated value/property...